llvm/lib/TargetParser/TargetParser.cpp

```diff
@@ -309,7 +309,45 @@ StringRef AMDGPU::getCanonicalArchName(const Triple &T, StringRef Arch) {
 void AMDGPU::fillAMDGPUFeatureMap(StringRef GPU, const Triple &T,
                                   StringMap<bool> &Features) {
   // XXX - What does the member GPU mean if device name string passed here?
-  if (T.isAMDGCN()) {
+  if (T.isSPIRV() && T.getOS() == Triple::OSType::AMDHSA) {
+    // AMDGCN SPIRV must support the union of all AMDGCN features.
+    Features["atomic-ds-pk-add-16-insts"] = true;
+    Features["atomic-flat-pk-add-16-insts"] = true;
+    Features["atomic-buffer-global-pk-add-f16-insts"] = true;
+    Features["atomic-global-pk-add-bf16-inst"] = true;
+    Features["atomic-fadd-rtn-insts"] = true;
+    Features["ci-insts"] = true;
+    Features["dot1-insts"] = true;
+    Features["dot2-insts"] = true;
+    Features["dot3-insts"] = true;
+    Features["dot4-insts"] = true;
+    Features["dot5-insts"] = true;
+    Features["dot7-insts"] = true;
+    Features["dot8-insts"] = true;
+    Features["dot9-insts"] = true;
+    Features["dot10-insts"] = true;
+    Features["dot11-insts"] = true;
+    Features["dl-insts"] = true;
+    Features["16-bit-insts"] = true;
+    Features["dpp"] = true;
+    Features["gfx8-insts"] = true;
+    Features["gfx9-insts"] = true;
+    Features["gfx90a-insts"] = true;
+    Features["gfx940-insts"] = true;
+    Features["gfx10-insts"] = true;
+    Features["gfx10-3-insts"] = true;
+    Features["gfx11-insts"] = true;
+    Features["gfx12-insts"] = true;
```

Comment:
What do these features even mean in the context of SPIR-V? It's basically a format for JIT, so can we really say stuff like GFX12 is available?

Reply:
One can neither say it's available nor say it's not, it's undecidable at SPIRV generation time (if we knew, we'd not need it in the first place and would've just targeted the actual hardware). The idea is that AMDGCN SPIRV will "support" the union of AMDGCN features, so that the associated builtins / constructs are retained in the generated SPIRV. Afterwards, at JIT time, one handles matching to the actual physical target that is present; this is done via a piece of paired functionality in the ME which has not yet been submitted.

Reply:
I see, so it's basically just pushing any target specific errors into when it's actually compiled to a binary.

Reply:
Yes, that is correct; the ME piece I mentioned will deal with avoiding spurious errors such as a piece of dead code invoking a `gfx12` instruction when JITing to a `gfx9` target, so that we only fail on actual misuses.